fontir/src/orchestration.rs

```diff
@@ -34,83 +32,306 @@ bitflags! {
+    }
+}
+
+impl<I, T, Ir> ContextMap<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: IdAware<I> + Persistable,
+    Ir: IrIo<I>,
+{
+    pub fn set_unconditionally(&self, value: T) {
+        let key = value.id();
+        self.acl.assert_write_access(&key);
+
+        if self.irio.emit_ir() {
+            let mut writer = self.irio.writer(&key);
+            value.write(&mut writer);
         }
-    };
+
+        self.value.write().insert(key, Arc::from(value));
+    }
 }
 
-impl Default for Flags {
-    /// Match the way gftools configures fontmake by default
-    fn default() -> Self {
-        Flags::EMIT_IR | Flags::PREFER_SIMPLE_GLYPHS
+impl<I, T, Ir> ContextMap<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: IdAware<I> + PartialEq + Persistable,
+    Ir: IrIo<I>,
+{
+    pub fn set(&self, value: T) {
+        let key = value.id();
+        self.acl.assert_write_access(&key);
+
+        // nop?
+        if self
+            .value
+            .read()
+            .get(&key)
+            .map(|arc| **arc == value)
+            .unwrap_or(false)
+        {
+            eprintln!("set_if_changed {:?}, NO CHANGE!", key);
+            return;
+        }
+
+        eprintln!("set_if_changed {:?}, changed!", key);
+        self.set_unconditionally(value);
     }
 }
 
+pub trait IdAware<I> {
+    fn id(&self) -> I;
+}
+
+pub trait Persistable {
+    fn read(from: &mut dyn Read) -> Self;
+    fn write(&self, to: &mut dyn Write);
+}
+
+/// Reads and writes IR
+pub trait IrIo<I> {
+    fn emit_ir(&self) -> bool;
+    /// None if there is nothing written down for id
+    fn reader(&self, id: &I) -> Option<Box<dyn Read>>;
+    fn writer(&self, id: &I) -> Box<dyn Write>;
+}
+
 // Unique identifier of work. If there are no fields work is unique.
 // Meant to be small and cheap to copy around.
 #[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
 pub enum WorkId {
     /// Build the initial static metadata.
-    InitStaticMetadata,
+    ///
+    /// Contains all names, axes, etc. Does NOT contain the final glyph order.
+    StaticMetadata,
     /// Build potentially variable font-wide metrics.
     GlobalMetrics,
     Glyph(GlyphName),
     GlyphIrDelete(GlyphName),
-    /// Update static metadata based on what we learned from IR
+    /// Glyph order from source, prior to adjustment
+    ///
+    /// Typically whatever2ir would populate this and then
+    /// it would be used to produce GlyphOrder.
     ///
-    /// Notably, IR glyphs with both components and paths may split into multiple
-    /// BE glyphs so the glyph order may change.
-    FinalizeStaticMetadata,
+    /// Most things should use GlyphOrder not this.
+    PreliminaryGlyphOrder,
+    /// The final glyph order. Most things that need glyph order should rely on this.
+    GlyphOrder,
     Features,
     Kerning,
 }
 
 pub type IrWork = dyn Work<Context, WorkId, WorkError> + Send;
 
+pub struct FeIrIo {
```

Comment:
All I see is *Fee-fi-fo-fum* 😬

Reply:
I was going to rename but now I don't want to!